### DISCUSSION TITLE
Generating Types with Typescript Template / String literal types

### DISCUSSION THREAD

**Discussion by kayvaninvemo** [mmkal/slonik-tools#348] [Ideas]
I was wondering if we can create something similar to the [this](https://github.com/codemix/ts-sql) in order to create a type safe `sql`.

The benefits I see is:

- generating types on the fly
- able to generate type for json_agg, json_build_object ...
- able to generate nullable fields for complicated queries 
- able to get the types from .join and .array
- able to generate types for nested queries (we have it on the issues)
- no need to changing the source code

But we need to connect to the database to read the table and column list

If you think the same, I can contribute some time for it, but I will need help.

**Comment by mmkal** [comment]
I'd be interested to see this in practice. I'm skeptical of how it'd perform, and whether it'd be able to handle parameterised queries like `` sql`select * from x where y = ${someValueFromJs}` `` - which are far more common than the examples in that repo which are just "plain" sql.
